### ISSUE TITLE
rearrange_ is sometimes called by only a subset of the communicator

### ISSUE THREAD

**Issue by worleyph** [E3SM-Project/E3SM#1979]
In a high resolution G case

     --res T62_oRRS18to6v3 --mach cori-knl

in which all components are stacked except ATM, which is given its own set of nodes, there are
calls to rearrange_ (in cime/src/externals/mct/mct/m_Rearrange.F90) that involve only the ATM processes even though the communicator used for the call includes all processes in both CPL and ATM. When this occurs, the MPI collective option (MPI_AlltoallV) hangs on Cori-KNL. swapm does not hang, because it does not require the involvement of processes that are not sending or receiving information (and apparently the Cori-KNL MPI library does). 

I am not sure if this is an issue with MCT or with the MPI library, or both. This does not happen in typical usage of E3SM because we currently specify swapm instead of MPI_AlltoallV when usealltoall = .true. I found this when trying to diagnose some peculiar performance issues associated with CPL in this G case.

I am in the process of testing with an ne30_oECv3_ICG resolution A_WCYCL1850S with CPL overlapped on OCN (and not overlapped with ATM) to see if the issue arises in this situation as well, and will report back when the job completes.

**Comment by worleyph** [comment]
Hi @rljacob , should I close this or leave it open? It is a bug if we ever want to use MPI collectives in rearrange_, and is dangerous in any case without some documentation pointing out that the communicator used in rearrange_ could be larger than the number of processes actually calling rearrange_.

Determining whether it is feasible to fix this requires someone with more understanding about the design of MCT than I have (or ever hope to have).

**Comment by rljacob** [comment]
Yes leave it open.

**Comment by worleyph** [comment]
@rljacob, I tried looking at this again. I was able to reproduce the issue using a B case (A_WCYCL) and by modifying the code in rearrange_ (in m_Rearranger.F90) to force it to use MPI_AlltoallV whenever usealltoall is .true. (by hardwiring useswapm to .false.). In all cases, the communicator used by rearrange_ was MPI_COMM_WORLD and it always hung unless 

a) CPL communicator has all processes in MPI_COMM_WORLD
b) ROF communicator has all processes in MPI_COMM_WORLD

If ROF communicator is smaller than MPI_COMM_WORLD, it hangs during ROF intialization (rearrange_ called in the timer "r_i:mosarti_mosart_init").

If ROF communicator has all of MPI_COMM_WORLD processes but CPL communicator does not, the code then hangs in 

     if (iamin_CPLID) then
        ...
        call component_init_aream(infodata, rof_c2_ocn, samegrid_ao, samegrid_al, samegrid_ro, samegrid_lg)
        ...
     endif

in cime_comp_mod.F90 (pretty sure at least). rearrange_ is also called during the run loop many times, and, based on a cursory glance,, the following logic looks typical:

       if (iamin_CPLID .and. (atm_c2_ocn .or. atm_c2_ice)) then
          ...
          call prep_ocn_calc_a2x_ox(timer='CPL:ocnpre1_atm2ocn')
          ...
       endif

with m_rearrange_ called within prep_ocn_calc_a2x_ox. So this will also hang unless the CPL communicator includes all of MPI_COMM_WORLD.

In summary, this behavior is baked into the current design of MCT and how it is used in E3SM (and CESM I assume). The solution I would suggest is to remove the option to call MPI_AlltoallV in rearrange_ . If you want to retain this option, you will need to determine how to introduce the actual communicator being used in each of these instances, and not expect MPI_COMM_WORLD to work (unless there is some MPI option to allow MPI_AlltoallV to work when not all processes in the communicator call the collective).

**Comment by worleyph** [comment]
Closing this here. This is an MCT issue, not an E3SM issue, in that it has to be fixed on the MCT side first.

**Comment by worleyph** [comment]
@rljacob , so ... the issue here is my poor memory. This problem was identified and diagnosed by Sheri Mickelson in 2010 (initially on Intrepid), where we determined that the Cray MPI library at the time did not have an issue when not all processes in a communicator called MPI_AlltoallV, but most of the other vendor MPIs did. This is actually what prompted adding the flow-control (swapm) usealltoall option.

My proposal still stands - since no MPI supports subsets of processes calling MPI_AlltoallV now, either this option should be removed from rearrange_ or MCT should add the ability to specific a particular subcommunicator and not use ThisMCTWorld%MCT_comm in rearrange_.

**Comment by rljacob** [comment]
Pat, can you open an issue in https://github.com/MCSclimate/MCT?  Or I could do it.

**Comment by worleyph** [comment]
> Pat, can you open an issue in https://github.com/MCSclimate/MCT? Or I could do it.

Done.

**Comment by amametjanov** [comment]
Pat, just curious, did you run into any issues if you pass a non-global (CPL-local) communicator to mct init?
in seq_comm_mct.F90
```
 552     call mct_world_init(ncomps, DRIVER_COMM, comms, comps)
```

**Comment by worleyph** [comment]
> Pat, just curious, did you run into any issues if you pass a non-global (CPL-local) communicator to mct init? in seq_comm_mct.F90

Hi @amametjanov , I have not done any experiments of this type. I don't know enough about MCT to know what things (like this) should be poked at.

**Comment by amametjanov** [comment]
Okay, thank you. PR MCSclimate/MCT#55 is up online for testing.
